File: site/lib/coverage.test.ts
import { describe, expect, test } from "vitest";
import {
  identityCoverage,
  professionalCoverage,
  careerCoverage,
  newsCoverage,
  evidenceCoverage,
  claimsCoverage,
  controversiesCoverage,
  sentimentCoverage,
  attentionCoverage,
  youtubeCoverage,
  redditCoverage,
  wikipediaCoverage,
  historicalCoverage,
  sourceDiversityCoverage,
  buildCoverageReport,
  summaryDimensions,
} from "./coverage";
import type { MediaLink, RealProfile, TrendPoint } from "./api";
import type { Controversy } from "./controversy";
import type { CareerIntelligence } from "./career";
import type { ProfessionalIdentity } from "./professional-identity";
import type { EvidenceItem } from "./evidence";
import type { Claim } from "./claims";

const EMPTY_PROFESSIONAL: ProfessionalIdentity = {
  primary: null,
  secondary: [],
  roles: [],
  industries: [],
  specializations: [],
  expertise: [],
  careerStatus: null,
  unresolved: [],
  empty: true,
};

function media(overrides: Partial<MediaLink> = {}): MediaLink {
  return {
    id: "m1",
    title: "A story",
    url: "https://reuters.com/1",
    source: "Reuters",
    type: "news",
    publishedAt: "2024-01-01",
    sentimentScore: -0.4,
    sentimentTag: "negative",
    ...overrides,
  };
}

function evidence(overrides: Partial<EvidenceItem> = {}): EvidenceItem {
  return {
    evidenceId: "media-1",
    sourceUrl: "https://reuters.com/1",
    sourceName: "Reuters",
    sourceType: "news",
    title: "A story",
    publicationDate: "2024-01-01",
    snippet: null,
    category: "news",
    relatedControversies: [],
    relatedToSentiment: false,
    duplicateCount: 1,
    independentSourceCount: 1,
    evidenceStrength: "limited",
    strengthReason: "x",
    ...overrides,
  };
}

function controversy(overrides: Partial<Controversy> = {}): Controversy {
  return {
    title: "Episode",
    summary: "summary",
    category: "Legal",
    severity: 3,
    status: "ongoing",
    year: 2024,
    sources: ["https://reuters.com/1"],
    ...overrides,
  };
}

function claim(overrides: Partial<Claim> = {}): Claim {
  return {
    claimId: "episode-base",
    entityId: null,
    controversyId: "episode",
    timelineEventId: null,
    claimText: "text",
    claimType: "reported_event",
    dateContext: null,
    status: "supported",
    confidence: "high",
    supportingEvidenceIds: ["media-1"],
    contradictingEvidenceIds: [],
    neutralEvidenceIds: [],
    responseEvidenceIds: [],
    createdAt: null,
    updatedAt: null,
    methodologyVersion: "cvm-1",
    statusReason: "reason",
    ...overrides,
  };
}

describe("identityCoverage", () => {
  test("no wikidata id -> unavailable", () => {
    expect(identityCoverage(undefined, "high").level).toBe("unavailable");
  });
  test("high resolution -> high, available", () => {
    const r = identityCoverage("Q1", "high");
    expect(r.level).toBe("high");
    expect(r.status).toBe("available");
  });
  test("ambiguous resolution -> insufficient, conflicting", () => {
    const r = identityCoverage("Q1", "ambiguous");
    expect(r.level).toBe("insufficient");
    expect(r.status).toBe("conflicting");
  });
});

describe("professionalCoverage", () => {
  test("empty -> unavailable", () => {
    expect(professionalCoverage(EMPTY_PROFESSIONAL).level).toBe("unavailable");
  });
});

describe("careerCoverage", () => {
  const base: CareerIntelligence = {
    timeline: [],
    organizations: [],
    insights: { start: null, current: null, transitions: [], leadershipRoles: [], founder: false, progression: [] },
    available: false,
  };

  test("no records -> unavailable", () => {
    expect(careerCoverage(base).level).toBe("unavailable");
  });

  test("case: strong entity resolution but weak news coverage — career can still read high independently", () => {
    const c: CareerIntelligence = {
      ...base,
      available: true,
      timeline: Array.from({ length: 6 }, (_, i) => ({
        start: 2010 + i,
        end: null,
        role: "Role",
        organization: "Org",
        location: null,
        industry: null,
        current: false,
        source: { name: "Wikidata", url: "https://www.wikidata.org/wiki/Q1" },
      })),
    };
    expect(careerCoverage(c).level).toBe("high");
  });
});

describe("newsCoverage — popular person with low evidence coverage vs less-famous person with high evidence coverage", () => {
  test("case 1: a popular person with only 2 duplicate-source articles reads low/limited", () => {
    const items = [media({ id: "1", source: "TMZ" }), media({ id: "2", source: "TMZ" })];
    const r = newsCoverage(items);
    expect(r.level).toBe("low");
    expect(r.status).toBe("limited");
  });

  test("case 2: a less-famous person with many independent reports reads high", () => {
    const items = Array.from({ length: 60 }, (_, i) =>
      media({ id: `${i}`, source: `Publisher ${i % 6}` }),
    );
    const r = newsCoverage(items);
    expect(r.level).toBe("high");
    expect(r.status).toBe("available");
  });

  test("case 3: a profile with only one source stays limited even with many articles", () => {
    const items = Array.from({ length: 40 }, (_, i) => media({ id: `${i}`, source: "OnlyOutlet" }));
    const r = newsCoverage(items);
    expect(r.status).toBe("limited");
  });

  test("no news items -> unavailable, never '0'", () => {
    const r = newsCoverage([]);
    expect(r.level).toBe("unavailable");
    expect(r.reasons.join(" ")).not.toMatch(/\b0\b/);
  });
});

describe("evidenceCoverage", () => {
  test("case 4: conflicting evidence is surfaced as a conflicting status, not averaged away", () => {
    const items = [
      evidence({ evidenceId: "a", evidenceStrength: "conflicting", sentimentTag: "negative" }),
      evidence({ evidenceId: "b", evidenceStrength: "conflicting", sentimentTag: "positive" }),
    ];
    expect(evidenceCoverage(items).status).toBe("conflicting");
  });

  test("no evidence -> unavailable", () => {
    expect(evidenceCoverage([]).level).toBe("unavailable");
  });
});

describe("claimsCoverage", () => {
  test("zero controversies -> not_applicable, never a bare zero", () => {
    const r = claimsCoverage([], 0);
    expect(r.status).toBe("not_applicable");
    expect(r.reasons[0]).not.toMatch(/^0$/);
  });

  test("conflicting claims present -> conflicting status", () => {
    const r = claimsCoverage([claim({ status: "conflicting" })], 1);
    expect(r.status).toBe("conflicting");
  });
});

describe("controversiesCoverage — case 10-ish: completely unavailable / clean-record honesty", () => {
  test("zero controversies never reads as 'no controversies exist'", () => {
    const r = controversiesCoverage([]);
    expect(r.reasons[0]).toBe("No supported controversy records are currently available.");
    expect(r.reasons.join(" ")).not.toMatch(/no controversies exist/i);
  });
});

describe("sentimentCoverage — case 9: strong news but weak sentiment sample", () => {
  test("small sample with low method agreement reads low/limited even with lots of news", () => {
    const r = sentimentCoverage({ sampleSize: 4, confidence: 0.2, trend: [], media: [] });
    expect(r.level).toBe("low");
  });

  test("no sample -> unavailable", () => {
    expect(sentimentCoverage({ sampleSize: null, confidence: 0.9, trend: [], media: [] }).level).toBe(
      "unavailable",
    );
  });
});

describe("attentionCoverage", () => {
  test("no series -> unavailable", () => {
    expect(attentionCoverage(null).level).toBe("unavailable");
  });
  test("healthy series -> high", () => {
    const series = Array.from({ length: 90 }, (_, i) => ({ date: `2024-01-${(i % 28) + 1}`, views: 10 }));
    const r = attentionCoverage({ source: "wikipedia", series, summary: null });
    expect(r.level).toBe("high");
  });
});

describe("youtube / reddit coverage — provider with zero items reads unavailable, not zero", () => {
  test("youtube unavailable", () => {
    const r = youtubeCoverage([media({ type: "news" })]);
    expect(r.level).toBe("unavailable");
    expect(r.reasons[0]).not.toMatch(/\b0\b/);
  });
  test("reddit with data reads available", () => {
    const items = Array.from({ length: 6 }, (_, i) =>
      media({ id: `${i}`, type: "reddit", source: `r/sub${i % 2}` }),
    );
    expect(redditCoverage(items).status).toBe("available");
  });
});

describe("wikipediaCoverage", () => {
  test("no summary/background/image -> unavailable", () => {
    expect(wikipediaCoverage({ summary: "", background: "" }).level).toBe("unavailable");
  });
  test("summary text present -> high", () => {
    expect(wikipediaCoverage({ summary: "Bio text.", background: "" }).level).toBe("high");
  });
});

describe("historicalCoverage — case 6 & 7: no snapshots vs sparse snapshots", () => {
  test("case 6: zero snapshots -> unavailable", () => {
    expect(historicalCoverage([]).level).toBe("unavailable");
  });

  test("case 7: sparse snapshots with a real gap are flagged, never invented as continuous", () => {
    const trend: TrendPoint[] = [
      { date: "2024-01-01", score: 0.1, mentions: 5 },
      { date: "2024-06-01", score: 0.1, mentions: 5 },
      { date: "2024-06-02", score: 0.1, mentions: 5 },
    ];
    const r = historicalCoverage(trend);
    expect(r.level).toBe("low");
    expect(r.timeRange?.gapNote).toBeTruthy();
  });
});

describe("sourceDiversityCoverage", () => {
  test("case 5: provider failure — zero media and zero evidence sources -> unavailable, not 0", () => {
    const r = sourceDiversityCoverage([], []);
    expect(r.level).toBe("unavailable");
    expect(r.reasons.join(" ")).not.toMatch(/\b0\b/);
  });
});

describe("coverage vs popularity", () => {
  test("news coverage never reads a popularity/follower field — only real article/publisher counts", () => {
    const r = newsCoverage([media({ source: "A" }), media({ id: "2", source: "B" })]);
    for (const reason of r.reasons) {
      expect(reason).not.toMatch(/follower|view count|upvote|trending/i);
    }
  });
});

describe("buildCoverageReport / summaryDimensions", () => {
  function fullProfile(): RealProfile {
    return {
      slug: "jane-doe",
      name: "Jane Doe",
      verified: true,
      wikidataId: "Q1",
      profession: "Executive",
      summary: "Jane Doe is an executive.",
      background: "",
      notableWorks: [],
      fetchedAt: new Date().toISOString(),
      sentimentScore: -0.2,
      trendDirection: "down",
      explanation: "",
      confidence: 0.8,
      scoreLow: null,
      scoreHigh: null,
      sampleSize: 428,
      methodAgreement: 0.8,
      positiveRatio: 0.2,
      neutralRatio: 0.3,
      negativeRatio: 0.5,
      positiveCount: 80,
      neutralCount: 120,
      negativeCount: 228,
      scoreNews: null,
      scoreYoutube: null,
      scoreInstagram: null,
      trend: [{ date: "2024-01-01", score: -0.2, mentions: 10 }],
      evidence: [],
      controversies: [controversy()],
      media: [media()],
      attention: { source: "wikipedia", series: [{ date: "2024-01-01", views: 100 }], summary: null },
      timeline: [],
      accounts: [],
      professional: EMPTY_PROFESSIONAL,
      career: {
        timeline: [],
        organizations: [],
        insights: { start: null, current: null, transitions: [], leadershipRoles: [], founder: false, progression: [] },
        available: false,
      },
      resolution: "high",
      candidates: [],
    };
  }

  test("produces every declared dimension exactly once", () => {
    const report = buildCoverageReport({ profile: fullProfile(), evidenceItems: [], claims: [] });
    const keys = report.dimensions.map((d) => d.key);
    expect(new Set(keys).size).toBe(keys.length);
    expect(report.coverageVersion).toBe("coverage-1");
  });

  test("summaryDimensions returns only the compact-card subset, in order", () => {
    const report = buildCoverageReport({ profile: fullProfile(), evidenceItems: [], claims: [] });
    const summary = summaryDimensions(report);
    expect(summary.map((d) => d.key)).toEqual([
      "identity",
      "career",
      "news",
      "evidence",
      "sentiment",
      "youtube",
      "historical",
    ]);
  });
});

File: test/coverage_test.dart
// The Data Coverage & Confidence Center's dimension-by-dimension
// calculations — never a single combined score, never a fabricated
// reason, never "0" standing in for "unavailable".
import 'package:flutter_test/flutter_test.dart';

import 'package:crititrack/core/domain/models/celebrity.dart';
import 'package:crititrack/core/domain/models/controversy.dart';
import 'package:crititrack/core/domain/models/media_item.dart';
import 'package:crititrack/core/domain/models/person_facts.dart';
import 'package:crititrack/core/domain/models/sentiment_data.dart';
import 'package:crititrack/core/utils/claims.dart';
import 'package:crititrack/core/utils/coverage.dart';
import 'package:crititrack/core/utils/evidence.dart';

MediaItem media({
  String id = '1',
  String title = 'A story',
  String? source = 'Reuters',
  MediaType type = MediaType.news,
  DateTime? publishedAt,
  String? channelTitle,
  String? sentimentTag,
}) => MediaItem(
  id: id,
  type: type,
  title: title,
  url: 'https://reuters.com/$id',
  source: source,
  publishedAt: publishedAt ?? DateTime.utc(2024, 1, 1),
  channelTitle: channelTitle,
  sentimentTag: sentimentTag,
);

EvidenceItem evidence({
  String evidenceId = 'media-1',
  EvidenceStrength evidenceStrength = EvidenceStrength.limited,
  String? sentimentTag,
}) => EvidenceItem(
  evidenceId: evidenceId,
  sourceUrl: 'https://reuters.com/1',
  sourceName: 'Reuters',
  sourceType: SourceType.news,
  title: 'A story',
  publicationDate: '2024-01-01',
  snippet: null,
  category: EvidenceCategory.news,
  relatedControversies: const [],
  relatedToSentiment: false,
  duplicateCount: 1,
  independentSourceCount: 1,
  evidenceStrength: evidenceStrength,
  strengthReason: 'x',
  sentimentTag: sentimentTag,
);

Controversy controversy() => const Controversy(
  title: 'Episode',
  summary: 'summary',
  category: ControversyCategory.legal,
  severity: 3,
  status: ControversyStatus.ongoing,
  year: 2024,
  sources: ['https://reuters.com/1'],
);

Claim claim({ClaimStatus status = ClaimStatus.supported}) => Claim(
  claimId: 'episode-base',
  entityId: null,
  controversyId: 'episode',
  timelineEventId: null,
  claimText: 'text',
  claimType: ClaimType.reportedEvent,
  dateContext: null,
  status: status,
  confidence: ClaimConfidence.high,
  supportingEvidenceIds: const ['media-1'],
  contradictingEvidenceIds: const [],
  neutralEvidenceIds: const [],
  responseEvidenceIds: const [],
  createdAt: null,
  updatedAt: null,
  methodologyVersion: 'cvm-1',
  statusReason: 'reason',
);

void main() {
  group('identityCoverage', () {
    test('no wikidata id -> unavailable', () {
      expect(identityCoverage(null, false).level, CoverageLevel.unavailable);
    });
    test('resolved and verified -> high, available', () {
      final r = identityCoverage('Q1', true);
      expect(r.level, CoverageLevel.high);
      expect(r.status, DataStatus.available);
    });
    test('resolved but unverified -> medium, limited (never fabricated confidence)', () {
      final r = identityCoverage('Q1', false);
      expect(r.level, CoverageLevel.medium);
      expect(r.status, DataStatus.limited);
    });
  });

  group('professionalCoverage', () {
    test('no occupations -> unavailable', () {
      expect(professionalCoverage(const []).level, CoverageLevel.unavailable);
    });
  });

  group('careerCoverage', () {
    test('no records -> unavailable', () {
      expect(careerCoverage(const []).level, CoverageLevel.unavailable);
    });

    test('case: strong entity resolution but weak news coverage — career reads independently high', () {
      final timeline = List.generate(
        6,
        (i) => CareerEntry(
          role: 'Role',
          organization: 'Org',
          start: 2010 + i,
          end: null,
          sourceName: 'Wikidata',
          sourceUrl: 'https://www.wikidata.org/wiki/Q1',
        ),
      );
      expect(careerCoverage(timeline).level, CoverageLevel.high);
    });
  });

  group('newsCoverage — case 1 & 2: popular vs. less-famous person', () {
    test('case 1: a popular person with only duplicate-source articles reads low/limited', () {
      final items = [media(id: '1', source: 'TMZ'), media(id: '2', source: 'TMZ')];
      final r = newsCoverage(items);
      expect(r.level, CoverageLevel.low);
      expect(r.status, DataStatus.limited);
    });

    test('case 2: a less-famous person with many independent reports reads high', () {
      final items = List.generate(
        60,
        (i) => media(id: '$i', source: 'Publisher ${i % 6}'),
      );
      final r = newsCoverage(items);
      expect(r.level, CoverageLevel.high);
      expect(r.status, DataStatus.available);
    });

    test('case 3: only one source stays limited even with many articles', () {
      final items = List.generate(40, (i) => media(id: '$i', source: 'OnlyOutlet'));
      expect(newsCoverage(items).status, DataStatus.limited);
    });

    test('no news -> unavailable, reasons never say a bare 0', () {
      final r = newsCoverage(const []);
      expect(r.level, CoverageLevel.unavailable);
      for (final reason in r.reasons) {
        expect(reason, isNot(matches(RegExp(r'\b0\b'))));
      }
    });
  });

  group('evidenceCoverage — case 4', () {
    test('conflicting evidence reads a conflicting status, never averaged away', () {
      final items = [
        evidence(evidenceId: 'a', evidenceStrength: EvidenceStrength.conflicting, sentimentTag: 'negative'),
        evidence(evidenceId: 'b', evidenceStrength: EvidenceStrength.conflicting, sentimentTag: 'positive'),
      ];
      expect(evidenceCoverage(items).status, DataStatus.conflicting);
    });

    test('no evidence -> unavailable', () {
      expect(evidenceCoverage(const []).level, CoverageLevel.unavailable);
    });
  });

  group('claimsCoverage', () {
    test('zero controversies -> not applicable, never a bare zero', () {
      final r = claimsCoverage(const [], 0);
      expect(r.status, DataStatus.notApplicable);
    });

    test('conflicting claims -> conflicting status', () {
      final r = claimsCoverage([claim(status: ClaimStatus.conflicting)], 1);
      expect(r.status, DataStatus.conflicting);
    });
  });

  group('controversiesCoverage — honesty about a clean record', () {
    test('zero controversies never reads as "no controversies exist"', () {
      final r = controversiesCoverage(const []);
      expect(r.reasons.first, 'No supported controversy records are currently available.');
    });
  });

  group('sentimentCoverage — case 9: strong news but weak sentiment sample', () {
    final base = SentimentData(
      overallScore: 40,
      positiveRatio: 0.2,
      negativeRatio: 0.5,
      neutralRatio: 0.3,
      trendDirection: 'down',
      explanation: '',
      trendData: const [],
      dominantEmotion: 'anger',
    );

    test('small sample with low confidence reads low', () {
      final data = base.copyWith(sampleSize: 4, confidence: 0.2);
      expect(sentimentCoverage(data, const []).level, CoverageLevel.low);
    });

    test('no sample -> unavailable', () {
      final data = base.copyWith(sampleSize: 0, confidence: 0.9);
      expect(sentimentCoverage(data, const []).level, CoverageLevel.unavailable);
    });
  });

  group('youtubeCoverage — zero items reads unavailable, not zero', () {
    test('unavailable when no videos retrieved', () {
      final r = youtubeCoverage([media(type: MediaType.news)]);
      expect(r.level, CoverageLevel.unavailable);
      for (final reason in r.reasons) {
        expect(reason, isNot(matches(RegExp(r'\b0\b'))));
      }
    });
  });

  group('wikipediaCoverage', () {
    test('no summary/background -> unavailable', () {
      const bio = Biography(profession: '', summary: '', background: '', notableWorks: [], controversies: []);
      expect(wikipediaCoverage(bio).level, CoverageLevel.unavailable);
    });
    test('summary text present -> high', () {
      const bio = Biography(
        profession: '',
        summary: 'Bio text.',
        background: '',
        notableWorks: [],
        controversies: [],
      );
      expect(wikipediaCoverage(bio).level, CoverageLevel.high);
    });
  });

  group('historicalCoverage — case 6 & 7: no snapshots vs. sparse snapshots', () {
    test('case 6: zero snapshots -> unavailable', () {
      expect(historicalCoverage(const []).level, CoverageLevel.unavailable);
    });

    test('case 7: sparse snapshots with a real gap are flagged, never invented as continuous', () {
      final trend = [
        SentimentSnapshot(
          date: '2024-01-01',
          positiveCount: 1,
          negativeCount: 1,
          neutralCount: 1,
          totalMentions: 3,
          dominantEmotion: 'neutral',
          score: 0.1,
        ),
        SentimentSnapshot(
          date: '2024-06-01',
          positiveCount: 1,
          negativeCount: 1,
          neutralCount: 1,
          totalMentions: 3,
          dominantEmotion: 'neutral',
          score: 0.1,
        ),
        SentimentSnapshot(
          date: '2024-06-02',
          positiveCount: 1,
          negativeCount: 1,
          neutralCount: 1,
          totalMentions: 3,
          dominantEmotion: 'neutral',
          score: 0.1,
        ),
      ];
      final r = historicalCoverage(trend);
      expect(r.level, CoverageLevel.low);
      expect(r.timeRange?.gapNote, isNotNull);
    });
  });

  group('sourceDiversityCoverage — case 5: provider failure', () {
    test('zero media and zero evidence sources -> unavailable, not 0', () {
      final r = sourceDiversityCoverage(const [], const []);
      expect(r.level, CoverageLevel.unavailable);
      for (final reason in r.reasons) {
        expect(reason, isNot(matches(RegExp(r'\b0\b'))));
      }
    });
  });

  group('coverage vs. popularity', () {
    test('news coverage reasons never mention followers/views/upvotes/trending', () {
      final r = newsCoverage([media(id: '1', source: 'A'), media(id: '2', source: 'B')]);
      for (final reason in r.reasons) {
        expect(reason.toLowerCase(), isNot(contains('follower')));
        expect(reason.toLowerCase(), isNot(contains('upvote')));
        expect(reason.toLowerCase(), isNot(contains('trending')));
      }
    });
  });

  group('buildCoverageReport / summaryDimensions', () {
    Celebrity fullCelebrity() => Celebrity(
      slug: 'jane-doe',
      name: 'Jane Doe',
      wikidataId: 'Q1',
      verified: true,
      biography: const Biography(
        profession: 'Executive',
        summary: 'Jane Doe is an executive.',
        background: '',
        notableWorks: [],
        controversies: [],
      ),
      sentimentData: SentimentData(
        overallScore: 40,
        positiveRatio: 0.2,
        negativeRatio: 0.5,
        neutralRatio: 0.3,
        trendDirection: 'down',
        explanation: '',
        trendData: const [],
        dominantEmotion: 'anger',
        sampleSize: 428,
        confidence: 0.8,
      ),
      mediaItems: [media()],
      fetchedAt: DateTime.utc(2024, 1, 1),
      facts: PersonFacts.empty,
    );

    test('produces every declared dimension exactly once', () {
      final report = buildCoverageReport(celebrity: fullCelebrity(), evidenceItems: const [], claims: const []);
      final keys = report.dimensions.map((d) => d.key).toSet();
      expect(keys.length, report.dimensions.length);
      expect(report.coverageVersion, 'coverage-1');
    });

    test('reddit is not among the produced dimensions (unsupported on mobile)', () {
      final report = buildCoverageReport(celebrity: fullCelebrity(), evidenceItems: const [], claims: const []);
      expect(
        report.dimensions.any((d) => d.label == 'Reddit'),
        isFalse,
      );
    });

    test('summaryDimensions returns only the compact-card subset, in order', () {
      final report = buildCoverageReport(celebrity: fullCelebrity(), evidenceItems: const [], claims: const []);
      final summary = summaryDimensions(report);
      expect(summary.map((d) => d.key).toList(), kSummaryDimensions);
    });
  });
}
